Upgrading to Actix 4.x

bendk opened the pull request:
## Description

Updating dependencies:
  - actix-web 4.x
  - actix-http 3.x
  - actix-rt 2.x
  - actix-cors 0.6.x
  - tokio 1.x
  - reqwest 0.11.x
  - deadpool 0.7.x

The unit tests are passing for me, but I'm not sure if this is ready to merge yet.  One reason I'm making a PR is that I think it will run the integration tests, which I gave up on getting to work on my machine.

I'm not sure how to land these changes.  I think some of these changes can be factored out and landed without upgrading any dependencies, I'll try to open a PR for that.  Other than that, I'm not sure if this can be split apart more.  I don't think it makes sense to upgrade any one of the dependencies without upgrading the others since that will result it duplicate crates getting compiled in (for example, the tests fail if you don't upgrade `reqwest` because the tokio runtimes get mixed up).

## Issue(s)

Closes #1249
Closes #1258

bendk commented on syncserver/src/web/extractors.rs:
```diff
@@ -134,7 +134,7 @@ impl From<BatchBsoBody> for PostCollectionBso {
 fn get_accepted(req: &HttpRequest, accepted: &[&str], default: &'static str) -> String {
     let mut candidates = Accept::parse(req).unwrap_or_else(|_| {
-        Accept(vec![qitem(
+        Accept(vec![QualityItem::max(
```
I got this from Ethan's doc, but please double check it because I have no clue what's actually going on here.

jrconlin replied:
Oh, wow. Yeah, this is a fairly convoluted way to deal with a very convoluted field specification. Basically, this is constructing the `Accept` header's `;q=` value, set to the max `1.0` for the Mime qualifier for whatever is specifed as `default`. (e.g. for `txt` it should produce something like `Accept: plain/txt:q=1.0`)

bendk commented on syncserver/src/server/mod.rs:
```diff
@@ -310,7 +314,7 @@ impl Server {
 
         if let Some(keep_alive) = actix_keep_alive {
-            server = server.keep_alive(keep_alive as usize);
+            server = server.keep_alive(Duration::from_secs(keep_alive as u64));
```
I'm pretty sure `keep_alive` is in seconds, but I'd love a double check on this one as well.

jrconlin replied:
Yes, actix-http 2.2.2 [used Seconds](https://docs.rs/actix-http/2.2.2/actix_http/enum.KeepAlive.html#variant.Timeout)

pjenvey replied:
The `keep_alive` method takes a `Duration` so the question's about if `settings.actix_keep_alive` is in seconds -- it is.

We usually document that in the Setting field's docstring but this one isn't documented -- let's clarify that it's seconds [here](https://github.com/mozilla-services/syncstorage-rs/blob/cc8d584/syncserver-settings/src/lib.rs)

bendk replied:
Good call, I just updated that.

bendk commented on syncserver/src/web/transaction.rs:
```diff
@@ -197,11 +195,10 @@ impl DbTransactionPool {
 
 impl FromRequest for DbTransactionPool {
-    type Error = Error;
+    type Error = actix_web::Error;
```
This was another change that felt kind of magical to me and I'd love some double-checking on.  This used to return `actix_http::Error`, which I think may have been a type alias for the `actix_web` version, but I'm not sure.  The reason the change is necessary, is `actix_http::Error` no longer implements `From<T: ResponseError>` after the upgrade.

jrconlin replied:
Yeah, this is pretty common for Actix, so I'm not surprised they moved `Error` again. Looks like the current version re-exports `actix_http::error::Error` which appears to be more about [HTTP processing errors](https://docs.rs/actix-http/3.5.1/src/actix_http/error.rs.html#81), rather than [`actix-web::error::Error`](https://docs.rs/actix-web/latest/actix_web/error/struct.Error.html), which appears to be very much about responses.

bendk commented on syncserver/src/server/test.rs:
```diff
@@ -149,7 +149,7 @@ fn create_request(
                 .expect("Could not get hn in create_request");
             let hv = HeaderValue::from_str(v.as_str()).expect("Could not get hv in create_request");
-            req = req.header(hn, hv);
+            req = req.insert_header((hn, hv));
```
The docs claim that `append_header` is equivalent to the old `header` method, but that wasn't true for `TestRequest`.  Here, we need to use `insert_header` to override the header set by `set_json` in order to make the `invalid_content_type` test pass.

Maybe we should keep the calls to `header` and update them in another pass, I'm not sure though.

jrconlin replied:
Yeah, the original [`header`](https://docs.rs/actix-web/3.3.3/src/actix_web/test.rs.html#461-469) just set the header in the request directly, and looking at the more recent code [`append_header`](https://docs.rs/actix-web/4.4.1/actix_web/test/struct.TestRequest.html#method.append_header) adds headers of the Key to the request, and [`insert_header`](https://docs.rs/actix-web/4.4.1/actix_web/test/struct.TestRequest.html#method.insert_header) replaces them.

I think we want to use `insert_header` for all of these since we want to create or replace any default header, not just add a value to it.

bendk replied:
Yeah, that makes sense to me too.  I'll update the code to use `insert_header`

bendk commented:
> I think some of these changes can be factored out and landed without upgrading any dependencies, I'll try to open a PR for that.

After testing it out some more, I don't think there's much than can be done before the upgrades.

jrconlin commented on syncserver/src/server/test.rs:
```diff
@@ -131,15 +135,15 @@ fn create_request(
     let mut req = test::TestRequest::with_uri(path)
         .method(method.clone())
-        .header(
+        .append_header((
```
```suggestion
        .insert_header((
```

jrconlin commented on syncserver/src/server/test.rs:
```diff
@@ -131,15 +135,15 @@ fn create_request(
-        .header("Accept", "application/json")
-        .header(
+        ))
+        .append_header(("Accept", "application/json"))
```
```suggestion
        .insert_header(("Accept", "application/json"))
```

jrconlin commented on syncserver/src/server/test.rs:
```diff
@@ -131,15 +135,15 @@ fn create_request(
-        .header(
+        ))
+        .append_header(("Accept", "application/json"))
+        .append_header((
```
```suggestion
        .insert_header((
```

jrconlin commented on syncserver/src/tokenserver/extractors.rs:
```diff
@@ -748,9 +740,9 @@ mod tests {
-            .header("authorization", "Bearer fake_token")
-            .header("accept", "application/json,text/plain:q=0.5")
-            .header("x-keyid", "0000000001234-qqo")
+            .append_header(("authorization", "Bearer fake_token"))
```
I think we're ok with using `append_header` in this case, since there's probably no existing header to add the value to, but we might want to swap this with `insert_header` to override any existing ones.

jrconlin reviewed (changes requested):
That's awesome that you were able to bump up tokio with this. That clears a lot of the CVEs in audit.toml as well. (We might want to do a separate commit that does a `cargo update` and then checks to see what we can purge out of  `.cargo/audit.toml`)

I think the only thing we might want to do is swap out most of the `.append_header()` calls for `.insert_header()`. My read of the docs and the source seems to imply that `.append_header()` adds the provided value to the list of values for that `key` where `.insert_header()` overwrites whatever was there before. Tests are passing because most of the time there's no prior header, but we probably should be paranoid about that. 
Thanks!

jrconlin reviewed (commented):
That's awesome that you were able to bump up tokio with this. That clears a lot of the CVEs in audit.toml as well. (We might want to do a separate commit that does a `cargo update` and then checks to see what we can purge out of  `.cargo/audit.toml`)

I think the only thing we might want to do is swap out most of the `.append_header()` calls for `.insert_header()`. My read of the docs and the source seems to imply that `.append_header()` adds the provided value to the list of values for that `key` where `.insert_header()` overwrites whatever was there before. Tests are passing because most of the time there's no prior header, but we probably should be paranoid about that. 
Thanks!

bendk commented:
Updated the code to use `insert_header`, ran `cargo update` and removed the audit ignores.  I'm starting to feel good about this one, although I don't have a good way to judge the risk.  What should be the next step?

jrconlin commented:
This looks like a fairly light lift. For heavier things, we'd want to run a load test to see if there were any large impacts to performance, but I don't think that's a concern here since the upgrades aren't impacting much of the code flow or design. (Mostly it's function name changes and a bit of ownership changes.)

If you're interested, the somewhat crufty load test script we have is [here](https://github.com/mozilla-services/syncstorage-loadtest), I'll reach out to Kat to see if there's a more recent script.

jrconlin commented:
*sigh* I really wish that github would flag unsigned commits earlier rather than as the very last check.
